### PR TITLE
feat(query) : Binary Join Many to One / One to Many with group left/right

### PR DESCRIPTION
**Pull request by TanviBhavsar**
**Pull Request checklist**

- [x] The commit(s) message(s) follows the contribution [guidelines](CONTRIBUTING.md) ?
- [x] Tests for the changes have been added (for bug fixes / features) ?
- [ ] Docs have been added / updated (for bug fixes / features) ?

Support for Binary Join Many to One / One to Many with group left/right

### CODE REVIEW

File: query/src/main/scala/filodb/query/exec/BinaryJoinExec.scala [hunk 1]
```diff
@@ -119,7 +118,17 @@ final case class BinaryJoinExec(id: String,
       result = if (onLabels.nonEmpty) result.filter(lv => onLabels.contains(lv._1)) // retain what is in onLabel list
                else result.filterNot(lv => ignoringLabels.contains(lv._1)) // remove the labels in ignoring label list
     }
-    // TODO handle one-to-many case
+
+    if (grouping.isDefined) {
```
Comment (vishramachandran) [inline query/src/main/scala/filodb/query/exec/BinaryJoinExec.scala]: should this be an else if ?

File: query/src/main/scala/filodb/query/exec/BinaryJoinExec.scala [hunk 2]
```diff
@@ -119,7 +118,17 @@ final case class BinaryJoinExec(id: String,
...
+      grouping.get.foreach { x =>
+        val labelVal = oneSideKey.labelValues.get(Utf8Str(x))
+        if (labelVal.isDefined && !labelVal.get.toString.equals("")) {
+          result += (Utf8Str(x) -> labelVal.get)
+        }
+        else
+          result -= Utf8Str(x)
+      }
```
Comment (vishramachandran) [inline query/src/main/scala/filodb/query/exec/BinaryJoinExec.scala]: result = grouping.map { x =>
...
}

File: query/src/main/scala/filodb/query/exec/BinaryJoinExec.scala [hunk 3]
```diff
@@ -119,7 +118,17 @@ final case class BinaryJoinExec(id: String,
...
-    // TODO handle one-to-many case
+
+    if (grouping.isDefined) {
+      grouping.get.foreach { x =>
+        val labelVal = oneSideKey.labelValues.get(Utf8Str(x))
+        if (labelVal.isDefined && !labelVal.get.toString.equals("")) {
+          result += (Utf8Str(x) -> labelVal.get)
+        }
+        else
+          result -= Utf8Str(x)
```
Comment (vishramachandran) [inline query/src/main/scala/filodb/query/exec/BinaryJoinExec.scala]: Lets discuss how we can simplify this. We shouldn't be adding and removing.

File: query/src/test/scala/filodb/query/exec/BinaryJoinGroupingSpec.scala [hunk 4]
```diff
@@ -0,0 +1,311 @@
...
+import org.scalatest.{FunSpec, Matchers}
+import org.scalatest.concurrent.ScalaFutures
+
+import filodb.core.metadata.Column.ColumnType
+import filodb.core.query._
+import filodb.memory.format.{RowReader, ZeroCopyUTF8String}
+import filodb.memory.format.ZeroCopyUTF8String._
+import filodb.query._
+
+class BinaryJoinGroupingSpec extends FunSpec with Matchers with ScalaFutures {
```
Comment (vishramachandran) [inline query/src/test/scala/filodb/query/exec/BinaryJoinGroupingSpec.scala]: Good list of test cases 👍

File: prometheus/src/test/scala/filodb/prometheus/parse/ParserSpec.scala [hunk 5]
```diff
@@ -317,7 +317,7 @@ class ParserSpec extends FunSpec with Matchers {
...
 
       "method_code:http_errors:rate5m / ignoring(code) group_left method:http_requests:rate5m" ->
-        "BinaryJoin(PeriodicSeries(RawSeries(IntervalSelector(1524855688000,1524855988000),List(ColumnFilter(__name__,Equals(method_code:http_errors:rate5m))),List()),1524855988000,1000000,1524855988000),DIV,OneToMany,PeriodicSeries(RawSeries(IntervalSelector(1524855688000,1524855988000),List(ColumnFilter(__name__,Equals(method:http_requests:rate5m))),List()),1524855988000,1000000,1524855988000),List(),List(code))",
+        "BinaryJoin(PeriodicSeries(RawSeries(IntervalSelector(1524855688000,1524855988000),List(ColumnFilter(__name__,Equals(method_code:http_errors:rate5m))),List()),1524855988000,1000000,1524855988000),DIV,OneToMany,PeriodicSeries(RawSeries(IntervalSelector(1524855688000,1524855988000),List(ColumnFilter(__name__,Equals(method:http_requests:rate5m))),List()),1524855988000,1000000,1524855988000),List(),List(code),Some(List()))",
```
[the remark this answers is not in the document]
Comment (TanviBhavsar) [reply prometheus/src/test/scala/filodb/prometheus/parse/ParserSpec.scala]: Added test case when group left has labels

File: query/src/main/scala/filodb/query/LogicalPlan.scala [hunk 6]
```diff
@@ -113,7 +113,7 @@ case class BinaryJoin(lhs: PeriodicSeriesPlan,
                       cardinality: Cardinality,
                       rhs: PeriodicSeriesPlan,
                       on: Seq[String] = Nil,
-                      ignoring: Seq[String] = Nil) extends PeriodicSeriesPlan
+                      ignoring: Seq[String] = Nil, grouping: Option[Seq[String]] = None) extends PeriodicSeriesPlan
```
[the remark this answers is not in the document]
Comment (TanviBhavsar) [reply query/src/main/scala/filodb/query/LogicalPlan.scala]: Changed to Seq(String)

File: query/src/test/scala/filodb/query/exec/BinaryJoinGroupingSpec.scala (hunk 4 above)
Comment (TanviBhavsar) [reply query/src/test/scala/filodb/query/exec/BinaryJoinGroupingSpec.scala]: Thank You!

File: query/src/main/scala/filodb/query/exec/BinaryJoinExec.scala [hunk 7]
```diff
@@ -118,8 +118,19 @@ final case class BinaryJoinExec(id: String,
...
       result = if (onLabels.nonEmpty) result.filter(lv => onLabels.contains(lv._1)) // retain what is in onLabel list
                else result.filterNot(lv => ignoringLabels.contains(lv._1)) // remove the labels in ignoring label list
+    } else if (cardinality == Cardinality.OneToMany || cardinality == Cardinality.ManyToOne) {
+      include.foreach { x =>
+          val labelVal = oneSideKey.labelValues.get(Utf8Str(x))
+          labelVal.foreach { v =>
+            if (v.toString.equals(""))
+              // If label value is empty do not propagate to result and
+              // also delete from result
+              result -= Utf8Str(x)
```
Comment (vishramachandran) [inline query/src/main/scala/filodb/query/exec/BinaryJoinExec.scala]: @TanviBhavsar Is result a mutable map here ?

**Comment by TanviBhavsar** [comment]
> Approved.
> 
> Can we add additional documentation about the following in the code:
> 
> * group_left
> * group_right
> * include ?
> 
> We need the document the 1. expected behavior 2. How we achieve it in our code.
> 
> This is so that it is easier for the next person looking at this code. This is especially important since the prom documentation is quite sparse on this topic.
> 
> Thanks!

I have added documentation

**Pull request merged**